plutus-contract-exe/src/Language/Plutus/Contract/Wallet.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE FlexibleContexts #-}
+-- | Balance  `UnbalancedTx` values using the
+--   wallet API
+module Language.Plutus.Contract.Wallet(
+      balanceWallet
+    , balanceTx
+    , handleTx
+    ) where
+
+import           Control.Lens
+import           Control.Monad                        ((>=>))
+import           Control.Monad.Except
+import           Data.Bifunctor                       (second)
+import           Data.Map                             (Map)
+import qualified Data.Map                             as Map
+import           Data.Maybe                           (fromMaybe)
+import qualified Data.Set                             as Set
+import           Language.Plutus.Contract.Transaction (UnbalancedTx)
+import qualified Language.Plutus.Contract.Transaction as T
+import           Ledger.Tx                            (Tx, TxOut, TxOutRef)
+import qualified Ledger.Tx                            as Tx
+import           Ledger.Value                         (Value)
+import qualified Ledger.Value                         as Value
+import           Wallet.API                           (MonadWallet, PubKey, WalletAPIError)
+import qualified Wallet.API                           as WAPI
+import qualified Wallet.Emulator                      as E
+
+-- | Balance an unbalanced transaction in a 'MonadWallet' context.
+balanceWallet
+    :: (WAPI.MonadWallet m)
+    => UnbalancedTx
+    -> m Tx
+balanceWallet utx = do
+    pk <- WAPI.ownPubKey
+    addr <- WAPI.watchedAddresses
+    let utxo = addr ^. at (Tx.pubKeyAddress pk) . to (fromMaybe mempty)
+    balanceTx utxo pk utx
+
+-- | Balance an unbalanced transaction by adding public key inputs
+--   and outputs
+balanceTx
+    :: ( MonadError WalletAPIError m )
+    => Map TxOutRef TxOut
+    -- ^ Unspent transaction outputs that may be used to balance the
+    --   left hand side (inputs) of the transaction.
+    -> PubKey
+    -- ^ Public key, used to balance the right hand side (outputs) of
+    --   the transaction.
+    -> UnbalancedTx
+    -- ^ The unbalanced transaction
+    -> m Tx
+balanceTx utxo pk tx =
+    let (neg, pos) = Value.split (T.computeBalance tx)
```

Comment:
Yeah, so I think this should just require the `Index`, rather than expecting the unbalanced transaction to have the values in it.